### DISCUSSION TITLE
How to process order without account

### DISCUSSION THREAD

**Discussion by sbkjarmul** [vuestorefront/prestashop#45] [Q&A]
Hi,

I would like to process order without account using VSF + prestashop. I got stuck.  
I can't find order object structure. I see there is solution for user with accounts implemented. 
How to create order object based on cart, shipping methods and payment method and how can I send it via make() function to accomplish order?

I hope I will find help here.

**Comment by samberrry** [comment]
Hi @sbkjarmul ,

If you are looking for a way for order processing without account, I think guest checkout is that feature. It's all about guest checkout in API. First try to implement guest checkout in API, and the rest is simple, remove auth from front side.

**Reply by sbkjarmul** [reply]
Hi @samberrry,

I don't understand how order is processing. There is a method 'make' which has payment option as argument, but how is it connected with order that user has in cart?

**Reply by sbkjarmul** [reply]
@samberrry 
How to remove auth? 
![image](https://user-images.githubusercontent.com/37276167/187070084-efbd55c6-9dfe-48ba-b9a4-a5171634bea9.png)

I'm sorry but I can't imagine how guest checkout should look like. Could you give me some tips?

**Reply by samberrry** [reply]
@sbkjarmul 
sure.

to remove auth check you should modify files located at this directory:
packages/theme/middleware

You may know that, in Ps, the difference between guest checkout and doing checkout with an account, is just filling password. So, the API is same as create account/register, but without a password. And the rest depends on your flow in UX. 

So, at checkout steps you can just ask people their name, lastname, get email and done.

**Reply by samberrry** [reply]
try to use latest version of API v2.4.6:
https://github.com/binshops/prestashop-rest

if you do not send password to register API, that would be considered as guest checkout.

**Reply by sbkjarmul** [reply]
@samberrry 
Thank you! I didn't know. 
I'm gonna update my codebase and will be back. I would like know how to implement custom payment method called 'Przelewy24' or 'BlueMedia' to my shop.

**Reply by sbkjarmul** [reply]
@samberrry I faced an issue with guest account or I do something wrong. I created a composable useGuestCheckout with public method guestCheckout({email, firstName, lastName) it works properly - users are visible in Prestashop panel. The problem is I can't add address or loadCustomer because user is not authenticated. 

guestCheckout:
![image](https://user-images.githubusercontent.com/37276167/187529655-115325eb-15e2-4f3f-bab8-8460ab2d7814.png)

addAddress: 
![image](https://user-images.githubusercontent.com/37276167/187530125-b926ade9-9705-4224-a71b-557f67d2e1a4.png)


<img width="575" alt="image" src="https://user-images.githubusercontent.com/37276167/187530056-13998bf4-453a-445c-a8c4-f73c1f6fa140.png">

What do I wrong?

**Reply by samberrry** [reply]
about the payment methods, you need to implement the logic, you can see the current APIs for bankwire and pay by check to get hint, the important thing is where you should place the validateOrder method. The rest is gateway's logic.

and about the address creation yes, you should get this error, as the address endpoint is protected. First go to address controller, and add public `$guestAllowed = true;` an instance variable, and try to extend the class from AbstractRESTController not AbstractAuthRESTController, same thing for other controllers that you may get error for auth.

Or try something like this, I pushed some code on this branch: 
https://github.com/binshops/prestashop-rest/tree/feat/guest-address

**Reply by sbkjarmul** [reply]
@samberrry Could you explain how to zip file correctly when I change the rest api module by myself? I had unzipped the file, made a changes for guest account and when I create .zip file from it using macOs archiver it doesn't work. (prestashop returns warning during module upload that file is not valid)

**Reply by samberrry** [reply]
you can also just update files. 
but be sure that within your zip you have a root directory named binshopsrest, same as the module name. See the original module to see the dir structure.

**Comment by sbkjarmul** [comment]
Thanks for answer I will try it tomorrow. As I understand I have to update
rest api module itself?

Could you tell me how to set attributes (color and size) to product in
cart? I’m trying to pass it through id_products_attribute but I don’t have
idea how this param should look like.

**Reply by samberrry** [reply]
try to update your module with the changes on the branch I mentioned. No need to do a fresh install, just a file replacement is okay.

The combination part is a little tricky in PrestaShop, however it's already implemented in VSF. If you select a product combination and add it to cart, it will be added. If you want to know how it works accurately, you can take a look at the old classic theme, it's very useful, we exactly followed that flow in VSF as well. How? in your PrestaShop default classic theme, navigate to the product detail page (on a product that has combination), open up you browser dev tool and see the network section, then try to change the attribute, and then check the requests. If you check postman we have the same parameters, indeed the API works like the default Ps. 
How to get id_product_attribute? in subsequent request (with group[x] set and refresh set to true), you will get id_product_attribute which will be used for cart update.

As we have implemented this process in VSF, you can also deep dive into code and see the process.
